textdb/textdb-dataflow/src/main/java/edu/uci/ics/textdb/dataflow/join/SimilarityJoinPredicate.java

```diff
@@ -0,0 +1,195 @@
+     *   payload:      payload_inner            payload_outer
+     *   
+     * result_schema,      result_tuple
+     *   _id:              new_random_id
+     *   inner_content:    "join"
+     *   outer_content:    "john"
+     *   spanList:         [(0, 4, "join", inner_content)  (0, 4, "john", outer_content)]
+     *   payload:          [payload_inner, payload_outer]
+     *   
+     */
+    public SimilarityJoinPredicate(String joinAttributeName, Double similarity) {
+        if (similarity > 1) {
+            similarity = 1.0;
+        } else if (similarity < 0) {
+            similarity = 0.0;
+        }
+        this.similarity = similarity;
+        this.joinAttributeName = joinAttributeName;
+    }
+
+    @Override
+    public String getIDAttributeName() {
+        return SchemaConstants._ID;
+    }
+
+    @Override
+    public String getJoinAttributeName() {
+        return joinAttributeName;
+    }
+    
+    @Override
+    public Schema generateOutputSchema(Schema innerOperatorSchema, Schema outerOperatorSchema) throws DataFlowException {
+        List<Attribute> outputAttributeList = new ArrayList<>();
+        
+        // add _ID field first
+        outputAttributeList.add(SchemaConstants._ID_ATTRIBUTE);
+        
+        for (Attribute attr : innerOperatorSchema.getAttributes()) {
+            String attrName = attr.getFieldName();
+            FieldType attrType = attr.getFieldType();
+            // ignore _id, spanList, and payload
+            if (attrName.equals(SchemaConstants._ID) || attrName.equals(SchemaConstants.SPAN_LIST) 
+                    || attrName.equals(SchemaConstants.PAYLOAD)) {
+                continue;
+            }
+            outputAttributeList.add(new Attribute(INNER_PREFIX + attrName, attrType));
+        }
+        for (Attribute attr : outerOperatorSchema.getAttributes()) {
+            String attrName = attr.getFieldName();
+            FieldType attrType = attr.getFieldType();
+            // ignore _id, spanList, and payload
+            if (attrName.equals(SchemaConstants._ID) || attrName.equals(SchemaConstants.SPAN_LIST) 
+                    || attrName.equals(SchemaConstants.PAYLOAD)) {
+                continue;
+            }
+            outputAttributeList.add(new Attribute(OUTER_PREFIX + attrName, attrType));
+        }
+        
+        // add spanList field
+        outputAttributeList.add(SchemaConstants.SPAN_LIST_ATTRIBUTE);
+
+        // add payload field if one of them contains payload
+        if (innerOperatorSchema.containsField(SchemaConstants.PAYLOAD) || 
+                outerOperatorSchema.containsField(SchemaConstants.PAYLOAD)) {
+            outputAttributeList.add(SchemaConstants.PAYLOAD_ATTRIBUTE);
+        }
+        
+        return new Schema(outputAttributeList.stream().toArray(Attribute[]::new));
+    }
+
+    @Override
+    public ITuple joinTuples(ITuple outerTuple, ITuple innerTuple, Schema outputSchema) throws DataFlowException {        
```

Comment:
Which one should be the first, `outerTuple` or `innerTuple`?  Function `generateOutputSchema ()` uses `inner` first.  Let's make them consistent in all functions.